File: VB/Editors.Win/LookUpEditEx.vb
Imports System.ComponentModel
Imports DevExpress.ExpressApp.Win.Core
Imports DevExpress.ExpressApp.Localization

Namespace Editors.Win
	<ToolboxItem(False)> _
	Public Class LookUpEditEx
		Inherits DevExpress.XtraEditors.LookUpEdit
		Implements IGridInplaceEdit
		Private gridEditingObject_Renamed As Object
		Shared Sub New()
			RepositoryItemLookUpEditEx.Register()
		End Sub
		Public Sub New()
			AddHandler MyBase.DataBindings.CollectionChanged, AddressOf DataBindings_CollectionChanged
		End Sub
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			If disposing Then
				RemoveHandler MyBase.DataBindings.CollectionChanged, AddressOf DataBindings_CollectionChanged
			End If
			MyBase.Dispose(disposing)
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemLookUpEditEx.EditorName
			End Get
		End Property
		Public Shadows ReadOnly Property Properties() As RepositoryItemLookUpEditEx
			Get
				Return CType(MyBase.Properties, RepositoryItemLookUpEditEx)
			End Get
		End Property
		Public Overrides Property EditValue() As Object
			Get
				Return MyBase.EditValue
			End Get
			Set(ByVal value As Object)
				If value IsNot DBNull.Value AndAlso value IsNot Nothing Then
					If Not Properties.Helper.LookupObjectType.IsInstanceOfType(value) Then
						If Properties.ThrowExceptionOnInvalidLookUpEditValueType Then
							Throw New InvalidCastException(SystemExceptionLocalizer.GetExceptionMessage(ExceptionId.UnableToCast, value.GetType(), Properties.Helper.LookupObjectType))
						Else
							MyBase.EditValue = Nothing
							Return
						End If
					End If
				End If
				MyBase.EditValue = value
			End Set
		End Property
		Public Function FindEditingObject() As Object
			Return BindingHelper.FindEditingObject(Me)
		End Function
		Private Sub OnEditingObjectChanged()
			If FindEditingObject() Is Nothing AndAlso EditValue IsNot Nothing Then
				EditValue = Nothing
			End If
		End Sub
		Private Sub DataBindings_CollectionChanged(ByVal sender As Object, ByVal e As CollectionChangeEventArgs)
			OnEditingObjectChanged()
		End Sub
		Public Shadows Sub UpdateDisplayText()
			MyBase.UpdateDisplayText()
			MyBase.Refresh()
		End Sub

		#Region "IGridInplaceEdit Members"
		Private ReadOnly Property IGridInplaceEdit_DataBindings() As ControlBindingsCollection Implements IGridInplaceEdit.DataBindings
			Get
				Return MyBase.DataBindings
			End Get
		End Property
		Private Property GridEditingObject() As Object Implements IGridInplaceEdit.GridEditingObject
			Get
				Return gridEditingObject_Renamed
			End Get
			Set(ByVal value As Object)
				If gridEditingObject_Renamed IsNot value Then
					gridEditingObject_Renamed = value
					OnEditingObjectChanged()
				End If
			End Set
		End Property
		#End Region
	End Class
End Namespace

File: CS/EFCore/Editors.Win/LookUpEditEx.cs
using System;
using System.ComponentModel;
using DevExpress.ExpressApp.Win.Core;
using DevExpress.ExpressApp.Localization;

namespace Editors.Win {
    [ToolboxItem(false)]
    public class LookUpEditEx : DevExpress.XtraEditors.LookUpEdit, IGridInplaceEdit {
        private object gridEditingObject;
        static LookUpEditEx() { RepositoryItemLookUpEditEx.Register(); }
        public LookUpEditEx() {
            base.DataBindings.CollectionChanged += DataBindings_CollectionChanged;
        }
        protected override void Dispose(bool disposing) {
            if(disposing) {
                base.DataBindings.CollectionChanged -= DataBindings_CollectionChanged;
            }
            base.Dispose(disposing);
        }
        public override string EditorTypeName {
            get { return RepositoryItemLookUpEditEx.EditorName; }
        }
        public new RepositoryItemLookUpEditEx Properties {
            get { return (RepositoryItemLookUpEditEx)base.Properties; }
        }
        public override object EditValue {
            get { return base.EditValue; }
            set {
                if(value != DBNull.Value && value != null) {
                    if(!Properties.Helper.LookupObjectType.IsInstanceOfType(value)) {
                        if(Properties.ThrowExceptionOnInvalidLookUpEditValueType) {
                            throw new InvalidCastException(SystemExceptionLocalizer.GetExceptionMessage(ExceptionId.UnableToCast,
                                value.GetType(),
                                Properties.Helper.LookupObjectType));
                        }
                        else {
                            base.EditValue = null;
                            return;
                        }
                    }
                }
                base.EditValue = value;
            }
        }
        public object FindEditingObject() {
            return BindingHelper.FindEditingObject(this);
        }
        private void OnEditingObjectChanged() {
            InitializeDataSource();
            if(FindEditingObject() == null && EditValue != null) {
                EditValue = null;
            }
        }
        public void InitializeDataSource() {
            if(Properties != null && Properties.Helper != null) {
                Properties.DataSource = Properties.Helper.CreateCollectionSource(FindEditingObject()).List;
            }
        }
        private void DataBindings_CollectionChanged(object sender, CollectionChangeEventArgs e) {
            OnEditingObjectChanged();
        }
        public new void UpdateDisplayText() {
            base.UpdateDisplayText();
            base.Refresh();
        }

        #region IGridInplaceEdit Members
        System.Windows.Forms.ControlBindingsCollection IGridInplaceEdit.DataBindings {
            get { return base.DataBindings; }
        }
        object IGridInplaceEdit.GridEditingObject {
            get { return gridEditingObject; }
            set {
                if(gridEditingObject != value) {
                    gridEditingObject = value;
                    OnEditingObjectChanged();
                }
            }
        }
        #endregion
    }
}
